### PR TITLE
Fixes the body part healing menu

### LINKED ISSUE #13545: Weird coloring in healing menu
Win 10, tiles, 0.C-3676.

To the left is actual health of my character. To the right is the healing menu popup. 

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26633815-weird-coloring-in-healing-menu?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by BevapDin**
Fixes #13545

It was slightly messed up because of the implicit assumption that the numeric values of the `body_part` and the `hp_part` enums match. E.g. it assumed that `(body_part)0 == bp_torso` and `(hp_part)0 == hp_head`, but `(hp_part)0` is actually `hp_head`.

Similar, the numbers 2 and 3 when casted to `body_part` give `bp_eyes` and `bp_mouth`, but when casted to `hp_part`, they give `hp_arm_l` and `hp_arm_r`. Neither `bp_eyes` and `bp_mouth` should be referenced here.

This is a screenshot from the current version:

screenschot-2015-09-12-14-39-36

The character has bites on left leg and left arm, infections on right leg and torso and is bleeding from head and both arms. You can see the mixed up colors. Each line should have the same color all over.

Same character, new code:

screenschot-2015-09-12-14-48-34

### CODE REVIEW

**Comment by vache** [comment]
This is just a suggestion, but if the item being used would cure the affliction (bandages for bleeding, first aid kit for infections, etc) should the right side reflect that?  So, head is bleeding and red |||\, and then you use a bandage, the first aid screen should show head: red |||\ -> green ||||.

**Comment by Night-Pryanik** [comment]
As far as I can tell, that was the original behavior not so long ago.

**Comment by BevapDin** [comment]
> This is just a suggestion, but if the item being used would cure the affliction (bandages for bleeding, first aid kit for infections, etc) should the right side reflect that? So, head is bleeding and red |||\, and then you use a bandage, the first aid screen should show head: red |||\ -> green ||||.

Yes, that's how it should probably be. I fixed that as well. Now the HP bars are in the limb color, the left bar is in the limb color before and the right bar is in the color after (successfully) using the item. The body part name is in the limb color that contains only the curable effects.

----

For example a rag (heals nothing, cures bleeding):

If the body part is only bleeding:
- print the name in red (color for bleeding)
- print the left HP in whatever it is in the UI (red)
- print the right HP in whatever it is without bleeding (gray)

If the body part is bleeding and infected:
- print the name in red (color for bleeding)
- print the left HP in whatever it is in the UI (yellow)
- print the right HP in whatever it is without bleeding (still infected, therefor green)

----

In short (effects can be bleeding, bite, or infected, no other effect counts):

- If the body part name is gray, no effect will be cured on it, but HP may get healed (or lost in case of cauterization).
- If the right HP bar is gray, all existing effects will be removed (if there are any).
- If the left HP bar and the body part name have a different color, only some (not all) of the effects will be removed.

**Comment by Coolthulhu** [comment]
Minor problem:
Healthy parts have green lifebars, just like infected parts, but disinfected parts have gray parts.

For example, using a first aid on a character with a damaged torso and infected leg will show:
(gray) torso (green) 99 -> (green) 100
(green) leg (green) 99 -> (gray) 100
Just dropping the coloration for healthy parts would work.

**Comment by Night-Pryanik** [comment]
Game still don't color damaged bodyparts in healing menu like they are in health interface.

**Comment by BevapDin** [comment]
The green color of the HP bars doesn't indicate anything (in this case), it's just a color, could as well be read or yellow or blue.

The HP bars and the body part names get colored if they have some effect (bite, bleed, infected):

- The name gets colored according to what effects it has and what can be cured (no overlap means gray, some curable effect: colored according to the curable effects).
- The left HP bar gets colored according to the current effects (including curable and non-curable).
- The right HP bar gets colored according to the current effects, minus the curable effects.

Green is for infected and red and yellow also indicate some effects. Coloring the HP bar according to the HP values (like in the sidebar) would be ambiguous: is it green because the HP are near maximum or is it green because the part is infected?

If the part has no effects at all, the color of the HP bar is simply green (for whatever reasons, it's just the default). This could be changed to gray, like the body part name?

**Comment by Coolthulhu** [comment]
> This could be changed to gray, like the body part name?

Yeah, if we aren't going all the way with r/g/b hp coloring, we may as well drop the ambiguity and give healthy parts a gray coloring and make it all obvious.

**Comment by Night-Pryanik** [comment]
Well, that was very convenient and informative (for me, at least) to see - in color - which my bodypart is most damaged. Color-marking of HP bars helps me comprehend it faster rather than only by symbolic representation.
How about that: special effects (if any) color only body part name, otherwise it's default color (gray), HP bars are colored like HP values? No ambiguity in this case - we see if we have special effects AND HP values.

**Comment by Coolthulhu** [comment]
Right side doesn't have part names, though. This would remove that extra bit of information: what effects does this healing item cure.

**Comment by Night-Pryanik** [comment]
Or maybe we can add body part name to the right, like "Torso (red color): || -> Torso (green color) |||"
Either way, I like the way it was not so long ago:

**Comment by Coolthulhu** [comment]
Adding extra torso sounds clunky. After all, it doesn't transform into anything.

I changed it from the old way because I wanted to convey that extra bit of information that you can't get from looking at the upper right corner of your screen.

Other ideas:
* Instead of (or in addition to) coloring the body part, add colored symbols (letters?) for effects that can be cured after the life bars. This would still work nicely if we had more curable effects (acid covering, swelling, burning etc.)
* Recolor effects so that they don't have the same colors as life bars, then have the life bars use the effect coloring first or hp coloring if no effect
* Try to cram the information into the arrow. For example, if the used item can't remove bleeding, color the arrow red.

**Comment by BevapDin** [comment]
> Right side doesn't have part names, though. This would remove that extra bit of information: what effects does this healing item cure.

The name is only colored with the curable effects. If the head is bitten and bleeding, and the torso is only bleeding, and one applies disinfectant (cures bites only), the head is blue (bitten) and the torso would still be gray.

Just coloring the bars according to the HP value can result in the same color *left* and *right*, not much information.

What about this: HP bars are gray if HP would not change (e.g. using a rag), otherwise the right HP bar would be green if HP are going to be healed or red if HP would be lost (e.g. when cauterizing).

**Comment by Night-Pryanik** [comment]
>Instead of (or in addition to) coloring the body part, add colored symbols (letters?) for effects that can >be cured after the life bars. This would still work nicely if we had more curable effects (acid covering, >swelling, burning etc.)

I like that. Example: torso is bleeding and bitten. "Torso || -> ||| (~~b~~ (red), ~~b~~ (blue)) 
Smth like this:
